File: yasmin_viewer/yasmin_viewer/yasmin_viewer_pub.py
# Copyright (C) 2023 Miguel Ángel González Santamarta
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import Any, Dict, List, Set
from rclpy.node import Node
import rclpy
import yasmin
from yasmin import StateMachine, State, Concurrence, OrthogonalState
from yasmin_ros.yasmin_node import YasminNode
from yasmin_msgs.msg import (
    State as StateMsg,
    StateMachine as StateMachineMsg,
    Transition as TransitionMsg,
)


class YasminViewerPub(object):
    """
    A class to publish the state of a Finite State Machine (FSM) for visualization.
    """

    def __init__(
        self,
        fsm: StateMachine,
        fsm_name: str = "",
        rate: int = 4,
        node: Node = None,
    ) -> None:
        """
        Initializes the YasminViewerPub instance.

        Args:
            fsm_name (str): The name of the FSM.
            fsm (StateMachine): The FSM instance to be published.
            rate (int): The rate in Hz at which to publish updates. Defaults to 4.
            node (Node, optional): A custom Node instance. If None, a new YasminNode is created.

        Raises:
            ValueError: If fsm_name is empty.
        """

        ## The ROS 2 node instance used for publishing.
        self._node = node

        if self._node is None:
            self._node: Node = YasminNode.get_instance()

        ## The name of the finite state machine.
        self._fsm_name: str = fsm_name

        if not self._fsm_name and not fsm.get_name():
            self._fsm_name = "Unnamed_FSM"
        elif not self._fsm_name:
            self._fsm_name = fsm.get_name()

        ## The finite state machine to be published.
        self._fsm: StateMachine = fsm

        ## The publisher for the state machine messages.
        self._pub = self._node.create_publisher(StateMachineMsg, "/fsm_viewer", 10)

        ## A timer to periodically publish the FSM state.
        self._timer = self._node.create_timer(1 / rate, self._publish_data)

    def shutdown(self) -> None:
        if self._timer is not None:
            self._timer.cancel()
            self._node.destroy_timer(self._timer)
            self._timer = None

        if self._pub is not None:
            self._node.destroy_publisher(self._pub)
            self._pub = None

    def parse_transitions(self, transitions: Dict[str, str]) -> List[TransitionMsg]:
        """
        Converts a dictionary of transitions into a list of TransitionMsg.

        Args:
            transitions (Dict[str, str]): A dictionary where keys are outcome names and values are state names.

        Returns:
            List[TransitionMsg]: A list of TransitionMsg representing the FSM transitions.
        """
        transitions_list = []

        for key in transitions:
            transition = TransitionMsg()
            transition.outcome = key
            transition.state = transitions[key]
            transitions_list.append(transition)

        return transitions_list

    def parse_state(
        self,
        state_name: str,
        state_info: Dict[str, Any],
        states_list: List[StateMsg],
        parent: int = -1,
    ) -> None:
        """
        Recursively parses a state and its transitions, adding the resulting StateMsg to the states list.

        Args:
            state_name (str): The name of the state.
            state_info (Dict[str, str]): Information about the state, including its transitions.
            states_list (List[StateMsg]): A list to which the parsed StateMsg will be appended.
            parent (int, optional): The ID of the parent state. Defaults to -1.

        Returns:
            None
        """
        state_msg = StateMsg()

        state_msg.id = len(states_list)
        state_msg.parent = parent

        # Extract state information
        state: State = state_info["state"]
        state_msg.name = state_name
        state_msg.transitions = self.parse_transitions(state_info["transitions"])
        state_msg.outcomes = state.get_outcomes()

        # Check if the state is a FSM, Concurrence, or OrthogonalState
        state_msg.is_fsm = isinstance(state, (StateMachine, Concurrence, OrthogonalState))

        # Add state to the list
        states_list.append(state_msg)

        # Parse child states if this state is a FSM
        if isinstance(state, StateMachine):
            fsm: StateMachine = state
            states = fsm.get_states()

            for child_state_name in states:
                child_state_info = states[child_state_name]
                self.parse_state(
                    child_state_name, child_state_info, states_list, state_msg.id
                )

            current_state = fsm.get_current_state()

            # Find the current state among the child states
            for child_state in states_list:
                if (
                    child_state.name == current_state
                    and child_state.parent == state_msg.id
                ):
                    state_msg.current_state = child_state.id
                    break

        # Parse child states if this state is a Concurrence
        elif isinstance(state, Concurrence):
            concurrence: Concurrence = state
            concurrent_states = concurrence.get_states()
            transitions = self.parse_concurrence_transitions(concurrence)
            state_msg.current_state = -2

            for child_state_name, child_state in concurrent_states.items():
                child_state_info = {
                    "state": child_state,
                    "transitions": {},
                }
                self.parse_state(
                    child_state_name, child_state_info, states_list, state_msg.id
                )
                states_list[-1].transitions = transitions[child_state_name]

                # Check if the child_state outcomes are in the transitions
                existing_outcomes: Set[str] = {
                    t.outcome for t in states_list[-1].transitions
                }
                for outcome in child_state.get_outcomes():
                    if outcome not in existing_outcomes:
                        # If not, add a transition to the default outcome of the concurrence
                        msg = TransitionMsg()
                        msg.outcome = outcome
                        msg.state = concurrence.get_default_outcome()
                        states_list[-1].transitions.append(msg)

        # Parse child states if this state is an OrthogonalState
        elif isinstance(state, OrthogonalState):
            ort: OrthogonalState = state
            regions = ort.get_regions()
            state_msg.current_state = -2

            for region in regions:
                region_msg = StateMsg()
                region_msg.id = len(states_list)
                region_msg.parent = state_msg.id
                region_msg.name = region.name
                region_msg.is_fsm = True
                region_msg.current_state = -1
                region_msg.outcomes = list(region.sm.get_outcomes())
                states_list.append(region_msg)

                region_states = region.sm.get_states()

                for child_state_name in region_states:
                    child_state_info = region_states[child_state_name]
                    self.parse_state(
                        child_state_name,
                        child_state_info,
                        states_list,
                        region_msg.id,
                    )

                region_current = region.sm.get_current_state()
                if region_current:
                    for child_state in states_list:
                        if (
                            child_state.name == region_current
                            and child_state.parent == region_msg.id
                        ):
                            region_msg.current_state = child_state.id
                            break

    def parse_concurrence_transitions(
        self, concurrence: Concurrence
    ) -> Dict[str, List[TransitionMsg]]:
        """
        Converts a concurrence outcome map into transition-like information for visualization.

        In concurrence, transitions are based on outcome combinations rather than direct state-to-state transitions.
        This method creates pseudo-transitions that represent the outcome mapping logic.

        Args:
            concurrence (Concurrence): The concurrence state to parse transitions from.

        Returns:
            Dict[str, List[TransitionMsg]]: Transition mappings per state name for visualization.
        """
        transitions = {}
        outcome_map = concurrence.get_outcome_map()

        # Add transitions for each outcome in the outcome map
        for outcome, requirements in outcome_map.items():
            for state_name, state_outcome in requirements.items():

                if state_name not in transitions:
                    transitions[state_name] = []
                msg = TransitionMsg()
                msg.outcome = state_outcome
                msg.state = outcome
                transitions[state_name].append(msg)

        return transitions

    def _publish_data(self) -> None:
        """
        Publishes the current state of the FSM.

        This method validates the FSM, gathers its state data, and publishes it.
        If validation fails, an error message is logged.

        Returns:
            None

        Raises:
            Exception: If the FSM validation fails, an error is logged and the function exits without publishing.
        """
        if not rclpy.ok():
            return

        try:
            self._fsm.validate()

            states_list = []
            self.parse_state(
                self._fsm_name,
                {"state": self._fsm, "transitions": {}},
                states_list,
            )

            state_machine_msg = StateMachineMsg()
            state_machine_msg.states = states_list

            yasmin.YASMIN_LOG_DEBUG(
                f"Publishing data of state machine '{self._fsm_name}'"
            )
            self._pub.publish(state_machine_msg)

        except ValueError as e:
            yasmin.YASMIN_LOG_ERROR(
                f"Not publishing state machine '{self._fsm_name}' due to validation failure: {e}"
            )

File: yasmin_viewer/src/yasmin_viewer/yasmin_viewer_pub.cpp
// Copyright (C) 2023 Miguel Ángel González Santamarta
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "yasmin_viewer/yasmin_viewer_pub.hpp"

#include "yasmin/concurrence.hpp"
#include "yasmin/logs.hpp"
#include "yasmin/types.hpp"
#include "yasmin_ros/yasmin_node.hpp"

using namespace yasmin_viewer;
using namespace std::chrono_literals;

YasminViewerPub::YasminViewerPub(yasmin::StateMachine::SharedPtr fsm)
    : YasminViewerPub(nullptr, fsm, "") {}

YasminViewerPub::YasminViewerPub(const rclcpp::Node::SharedPtr &node,
                                 yasmin::StateMachine::SharedPtr fsm)
    : YasminViewerPub(node, fsm, "") {}

YasminViewerPub::YasminViewerPub(yasmin::StateMachine::SharedPtr fsm,
                                 const std::string &fsm_name)
    : YasminViewerPub(nullptr, fsm, fsm_name) {}

YasminViewerPub::YasminViewerPub(const rclcpp::Node::SharedPtr &node,
                                 yasmin::StateMachine::SharedPtr fsm,
                                 const std::string &fsm_name)
    : fsm(fsm), fsm_name(fsm_name) {

  if (node == nullptr) {
    this->node_ = yasmin_ros::YasminNode::get_instance();
  } else {
    this->node_ = node;
  }

  if (this->fsm_name.empty() && this->fsm->get_name().empty()) {
    this->fsm_name = "Unnamed_FSM";
  } else if (this->fsm_name.empty()) {
    this->fsm_name = this->fsm->get_name();
  }

  this->publisher =
      this->node_->create_publisher<yasmin_msgs::msg::StateMachine>(
          "/fsm_viewer", 10);

  this->timer = this->node_->create_wall_timer(
      250ms, std::bind(&YasminViewerPub::publish_data, this));
}

std::vector<yasmin_msgs::msg::Transition> YasminViewerPub::parse_transitions(
    const yasmin::Transitions &transitions) const {
  std::vector<yasmin_msgs::msg::Transition> transitions_list;
  transitions_list.reserve(transitions.size());

  for (const auto &transition : transitions) {
    auto &transition_msg = transitions_list.emplace_back();
    transition_msg.outcome = transition.first;
    transition_msg.state = transition.second;
  }
  return transitions_list;
}

std::unordered_map<std::string, std::vector<yasmin_msgs::msg::Transition>>
YasminViewerPub::parse_concurrence_transitions(
    yasmin::Concurrence::SharedPtr concurrence) const {
  std::unordered_map<std::string, std::vector<yasmin_msgs::msg::Transition>>
      transitions;
  const auto &outcome_map = concurrence->get_outcome_map();

  // Add transitions for each outcome in the outcome map
  for (const auto &[outcome, requirements] : outcome_map) {
    for (const auto &[state_name, state_outcome] : requirements) {
      auto &transition_msg = transitions[state_name].emplace_back();
      transition_msg.outcome = state_outcome;
      transition_msg.state = outcome;
    }
  }

  return transitions;
}

void YasminViewerPub::parse_state(
    const std::string &state_name, yasmin::State::SharedPtr state,
    const yasmin::Transitions &transitions,
    std::vector<yasmin_msgs::msg::State> &states_list, int parent) {

  auto state_msg = yasmin_msgs::msg::State();

  state_msg.id = states_list.size();
  state_msg.parent = parent;

  // State information
  state_msg.name = state_name;
  state_msg.transitions = this->parse_transitions(transitions);
  auto outcomes = state->get_outcomes();
  state_msg.outcomes =
      std::vector<std::string>(outcomes.begin(), outcomes.end());

  // Check if state is a nested FSM, Concurrence, or OrthogonalState
  auto fsm = std::dynamic_pointer_cast<yasmin::StateMachine>(state);
  auto concurrence = std::dynamic_pointer_cast<yasmin::Concurrence>(state);
  auto orthogonal = std::dynamic_pointer_cast<yasmin::OrthogonalState>(state);
  state_msg.is_fsm =
      (fsm != nullptr) || (concurrence != nullptr) || (orthogonal != nullptr);

  // Add the state to the list
  states_list.push_back(state_msg);

  // Handle nested FSM states
  if (fsm != nullptr) {
    auto states = fsm->get_states();
    auto aux_transitions = fsm->get_transitions();

    for (const auto &nested_state : states) {
      this->parse_state(nested_state.first, nested_state.second,
                        aux_transitions[nested_state.first], states_list,
                        state_msg.id);
    }

    // Identify the current state in the FSM
    std::string current_state = fsm->get_current_state();
    for (auto &child_state : states_list) {
      if (child_state.name == current_state &&
          child_state.parent == state_msg.id) {
        states_list[state_msg.id].current_state = child_state.id;
        break;
      }
    }
  }
  // Handle concurrence states
  else if (concurrence != nullptr) {
    auto concurrent_states = concurrence->get_states();
    auto transitions = this->parse_concurrence_transitions(concurrence);
    states_list[state_msg.id].current_state =
        -2; // Special marker for concurrence

    for (const auto &[child_state_name, child_state] : concurrent_states) {
      yasmin::Transitions empty_transitions;
      this->parse_state(child_state_name, child_state, empty_transitions,
                        states_list, state_msg.id);

      // Set transitions for this child state
      if (transitions.find(child_state_name) != transitions.end()) {
        states_list.back().transitions = transitions[child_state_name];
      }

      // Check if the child_state outcomes are in the transitions
      auto child_outcomes = child_state->get_outcomes();
      std::set<std::string> transition_outcomes;
      for (const auto &transition : states_list.back().transitions) {
        transition_outcomes.insert(transition.outcome);
      }

      for (const auto &outcome : child_outcomes) {
        if (transition_outcomes.find(outcome) == transition_outcomes.end()) {
          // If not, add a transition to the default outcome of the concurrence
          auto msg = yasmin_msgs::msg::Transition();
          msg.outcome = outcome;
          msg.state = concurrence->get_default_outcome();
          states_list.back().transitions.push_back(msg);
        }
      }
    }
  }
  // Handle orthogonal states
  else if (orthogonal != nullptr) {
    const auto &regions = orthogonal->get_regions();
    states_list[state_msg.id].current_state = -2;

    for (const auto &region : regions) {
      // Create a synthetic container node for the region
      auto region_msg = yasmin_msgs::msg::State();
      region_msg.id = states_list.size();
      region_msg.parent = state_msg.id;
      region_msg.name = region.name;
      region_msg.is_fsm = true;
      region_msg.current_state = -1;
      auto region_outcomes = region.sm->get_outcomes();
      region_msg.outcomes = std::vector<std::string>(region_outcomes.begin(),
                                                     region_outcomes.end());
      states_list.push_back(region_msg);

      auto region_states = region.sm->get_states();
      auto region_transitions = region.sm->get_transitions();

      for (const auto &nested_state : region_states) {
        yasmin::Transitions transitions;
        auto trans_it = region_transitions.find(nested_state.first);
        if (trans_it != region_transitions.end()) {
          transitions = trans_it->second;
        }

        this->parse_state(nested_state.first, nested_state.second, transitions,
                          states_list, region_msg.id);
      }

      std::string region_current = region.sm->get_current_state();
      if (!region_current.empty()) {
        for (auto &child : states_list) {
          if (child.name == region_current && child.parent == region_msg.id) {
            states_list[region_msg.id].current_state = child.id;
            break;
          }
        }
      }
    }
  }
}

void YasminViewerPub::publish_data() {

  if (!rclcpp::ok()) {
    return;
  }

  try {
    this->fsm->validate();

    std::vector<yasmin_msgs::msg::State> states_list;
    this->parse_state(this->fsm_name, this->fsm, {}, states_list, -1);

    auto state_machine_msg = yasmin_msgs::msg::StateMachine();
    state_machine_msg.states = states_list;

    YASMIN_LOG_DEBUG("Publishing data of state machine '%s'",
                     this->fsm_name.c_str());
    this->publisher->publish(state_machine_msg);

  } catch (const std::exception &e) {
    YASMIN_LOG_ERROR(
        "Not publishing state machine '%s' due to validation failure: \"%s\"",
        this->fsm_name.c_str(), std::string(e.what()).c_str());
  }
}
